Features/bsh/25.2/js preferences

bschwarzent opened the pull request.

bschwarzent committed:
ObjectIdProvider: add support for uuidParent

When the uuidPath is computed, the UUID of the object is combined with
the UUID of its parent. However, there are some cases where the parent
is only set due to the DOM structure, while the "logical parent" is
something else (e.g. detail table of a page). This can now optionally
be specified by setting the "uuidParent" property.

bschwarzent committed:
TableOrganizerForm: various improvements

- Cleanup code
- Edit columns with double click

bschwarzent committed:
Table: add mappers between user filters and user filter state DOs

389574

bschwarzent committed:
UiPreferences: add support to store table preferences for Scout JS

Certain UI adjustments the user can make should be preserved between
sessions and devices. This can be done by storing those settings in
a user-specific data structure called "(Client) UI Preferences". In
Scout Classic, this feature is already present. This commit adds basic
support for pure Scout JS elements.

Currently, only table preferences are supported (e.g. column widths),
but the data structure can be extended in the future. To enable the
preservation of preferences for a table, assign it a 'uuid' and set
the new property 'uiPreferencesEnabled' to true. By default, the
feature is enabled only for detail tables of PageWithTable.

The preferences are saved to a "UiPreferenceStore". The default
implementation uses the browser's local storage. A custom implementation
can be provided via object factory.

389574

cguglielmo commented on eclipse-scout-core/src/prefs/UiPreferences.ts:
```diff
@@ -0,0 +1,494 @@
+   * **Important:** This method clears internal data structures, but does *not* automatically reload preferences
+   * from the new store. To do so, {@link load} has to be called manually.
+   *
+   * @return the old store
```
@return -> @returns

cguglielmo commented on eclipse-scout-core/src/prefs/UiPreferences.ts:
```diff
@@ -0,0 +1,494 @@
+    this._storeTimeoutId = setTimeout(() => {
+      this.store()
+        .catch(error => {
+          console.warn('Unable to store UI preferences', error);
```
don't use console logs, errorHandler can log it using logError

bschwarzent replied:
The idea was to print a meaningful message instead of only the original error (much might just be a network error without any context). But I agree that outputting to the browser console is often not that useful. I removed it.

cguglielmo commented on eclipse-scout-core/src/prefs/UiPreferences.ts:
```diff
@@ -0,0 +1,494 @@
+  }
+
+  /**
+   * If the table is customizable, return sthe customizer data. Otherwise, `null` is returned.
```
return sthe -> returns the

cguglielmo commented on eclipse-scout-core/src/prefs/UiPreferences.ts:
```diff
@@ -0,0 +1,494 @@
+    let prefs = this.getOrCreateTablePreferences(table);
+    let existingProfile = this.getTablePreferencesProfile(prefs, profileId);
+    if (existingProfile) {
+      if (objects.equals(profile, existingProfile)) {
```
I think profile.equals should be preferred

cguglielmo commented on eclipse-scout-core/src/table/organizer/TableOrganizer.ts:
```diff
@@ -208,18 +210,18 @@ export class TableOrganizer implements ObjectWithType {
+    if (this.table.isCustomizable() && this.table.customizer.isCustomizable(column)) {
+      return this.table.customizer.modifyColumn(column);
     }
+    // NOP (currently not supported)
```
Is this comment necessary? It will never be supported I assume

bschwarzent replied:
Comment adjusted

cguglielmo commented on eclipse-scout-core/src/table/Table.ts:
```diff
@@ -6156,8 +6466,90 @@ export class Table extends Widget implements TableModel, Filterable<TableRow> {
+    this.setProperty('uiPreferencesEnabled', !!uiPreferencesEnabled);
+  }
+
+  _setUiPreferencesEnabled(uiPreferencesEnabled: boolean) {
```
missing protected modifier

cguglielmo commented on eclipse-scout-core/src/table/Table.ts:
```diff
@@ -6156,8 +6466,90 @@ export class Table extends Widget implements TableModel, Filterable<TableRow> {
+  applyColumnPreferences(columnPreferences: TableColumnClientUiPreferenceDo[]) {
+    let columnPreferencesMap = new Map(arrays.ensure(columnPreferences).map(pref => [pref.columnId, pref]));
+    let newColumns = this.columns
+      .filter(c => !c.guiOnly)
```
maybe map to \[column, prefs] first to simplify the sort and map functions? 
.map(column => \[column, columnPreferencesMap.get(column.buildUuid())] as \[Column<any>, TableColumnClientUiPreferenceDo])

bschwarzent replied:
I'm not sure if all those brackets make it easier to read. I added some comments instead. The column properties are now applied in a separate forEach loop. I think it's pretty readable now.

cguglielmo commented on eclipse-scout-core/src/data-objects.ts:
```diff
@@ -188,3 +188,21 @@ export class TextColumnUserFilterStateDo extends BaseDoEntity implements IUserFi
+@typeName('scout.UiPreferences')
+export class UiPreferencesDo extends BaseDoEntity {
+  tablePreferences?: TableClientUiPreferencesDo[];
+  // --- Add when necessary: ---
```
I would remove this uncommented code

cguglielmo commented on eclipse-scout-core/src/scout.ts:
```diff
@@ -73,6 +73,13 @@ export interface ObjectWithUuid {
   uuid: string;
 
+  /**
+   * Widget to be used instead of the default parent when computing the uuid path for the object.
```
Widget -> Object

cguglielmo commented on eclipse-scout-core/src/scout.ts:
```diff
@@ -118,6 +125,12 @@ export interface ObjectModelWithUuid<TObject = object> extends ObjectModel<TObje
    */
   uuid?: string;
+  /**
+   * Widget to be used instead of the default parent when computing the uuid path for the object.
```
Widget -> Object

cguglielmo commented on eclipse-scout-core/src/table/userfilter/UserFilterStateMapper.ts:
```diff
@@ -0,0 +1,46 @@
+    return null;
+  }
+
+  protected _createColumnFilter(table: Table, columnId: string, model: any): TFilter {
```
This method is not used, column filter state mappers should probably use it?

bschwarzent replied:
You are correct, that method is left over from a previous version. It didn't seem to make the individual mappers much simpler, so I reverted it but forget to delete this method. It is now removed.

cguglielmo commented on eclipse-scout-core/src/table/Table.ts:
```diff
@@ -4247,6 +4291,27 @@ export class Table extends Widget implements TableModel, Filterable<TableRow> {
+
+    // Update filters
+    this.setFilters(newFilters);
+    this.filter();
```
Calling this.filter() is not necessary, setFilters already does it

bschwarzent replied:
Hm, you seem to be correct. I one had a situation where just calling setFilters() would not apply them, but I can't reproduce it anymore. Therefore, I removed the filter() call, as suggested.

cguglielmo commented on eclipse-scout-core/src/table/Table.ts:
```diff
@@ -6156,8 +6466,90 @@ export class Table extends Widget implements TableModel, Filterable<TableRow> {
+    column.setWidth(columnPreferences.width);
+
+    // Properties without setter (changes will be applied later by _setColumns)
+    column.sortIndex = columnPreferences.sortOrder;
```
sortActive needs to be set as well, otherwise sorting won't be applied correctly

bschwarzent replied:
Ok, I expected this to be a "computed" property which is automatically set. It is now also set here, derived from `sortIndex >= 0`.

cguglielmo commented on eclipse-scout-core/src/table/Table.ts:
```diff
@@ -6156,8 +6466,90 @@ export class Table extends Widget implements TableModel, Filterable<TableRow> {
+    // Use setter for 'visible' property because it is a multidimensional property
+    column.setVisible(columnPreferences.visible, false); // parameter 'false' skips call of onColumnVisibilityChanged()
+    // FIXME bsh [js-preferences] this triggers an unnecessary change. As a matter of fact, _renderViewport is called too many times anyway!
+    column.setWidth(columnPreferences.width);
```
I think it is ok to just assign the width without using the setter, I don't think we need to trigger a resize event

cguglielmo commented on eclipse-scout-core/src/table/Table.ts:
```diff
@@ -6156,8 +6466,90 @@ export class Table extends Widget implements TableModel, Filterable<TableRow> {
+    column.grouped = columnPreferences.groupingActive;
+
+    if (column instanceof NumberColumn) {
+      column.setAggregationFunction(columnPreferences.aggregationFunctionId as NumberColumnAggregationFunction);
```
Maybe add comment that setter is required to update the aggregation and background effect functions (e..g column.aggrStart etc.)

bschwarzent committed:
Table: don't sort when table is empty

If there are no rows, Table#_sort doesn't have to do anything. This
is more efficient, and it also fixes a problem when _sort is called
on the detail table of a PageWithTable before the outline is fully
initialized. For example, if the breadcrumb mode is enabled, Tree#_init
expands a node which might lead to PageWithTable#_onTableRowOrderChanged
being executed before Outline#_init had a change to create the mediator.
